### PR TITLE
Add a class to perform GeneralDataStorage

### PR DESCRIPTION
**Pull request by jppelteret**
Relates to discussion in #7806

### CODE REVIEW

**Comment by jppelteret** [comment]
@luca-heltai I've started to strip the functionality out of `ScratchData` to form a generic data storage class (an updated version of `AnyData`, as we'd disucssed). I would love for this to offer serialization functionality if possible, as this would greatly help in storing QP data. I need to think about that part a little more (and write some tests, of course). 

I also added the `get_or_add_object_with_name()` function, and also split `add_[copy/reference]` into two variants, namely `add_unique_[copy/reference]` and `add_or_overwrite_[copy/reference]`. I'd be happy to discuss all changes from the prototype interface originally laid out in `ScratchData`.

If you feel so inclined, please feel free to push more additions onto this.

**Comment by luca-heltai** [comment]
Would you mind adding a default copy constructor?

**Comment by luca-heltai** [comment]
Since the original code was written by me, I'll let others comment. :-)

**Comment by jppelteret** [comment]
> Since the original code was written by me, I'll let others comment. :-)

Fair enough, but are you content with the changed interface? Does it suit your needs?

**Comment by jppelteret** [comment]
Also, do I need to add that reference to deal2kit here? Or that only for `ScratchData`?

**Comment by luca-heltai** [comment]
This is perfect for my need. If you could add the reference to the deal2lkit paper, it would be fair. I'll change the other PR to use this interface.

**Comment by jppelteret** [comment]
Great. I've added the citation, and have removed the serialisation function. This doesn't really seem to be possible when using `boost::any`, so maybe we can make a variant of this class that only accepts and stores serialisable class instances (or something like that).

File: include/deal.II/algorithms/general_data_storage.h [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+   * Store internally a copy of the given object. The copied object is
+   * owned by this class, and is accessible via reference through the
+   * get_object_with_name() method.
+   *
+   * This function ensures that no @p entry with the given @name is
+   * already stored by this class instance.
+   */
+  template <typename Type>
+  void
+  add_unique_copy(const Type &entry, const std::string &name);
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/algorithms/general_data_storage.h]: same for all of the other functions.

File: include/deal.II/algorithms/general_data_storage.h [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+  add_or_overwrite_reference(Type &entry, const std::string &name);
+
+  /**
+   * Return a reference to the object with given name. If the object does
+   * not exist, then the input @p arguments will be used to construct an object
+   * of the given @p Type.
+   */
+  template <typename Type, typename... Args>
+  Type &
+  get_or_add_object_with_name(const std::string &name, Args &... arguments);
```
Comment (bangerth) [inline include/deal.II/algorithms/general_data_storage.h]: You mean an object of type `Type` is generated by calling the constructor with the given set of `arguments`? Maybe you can make that clearer?

File: include/deal.II/algorithms/general_data_storage.h [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+  /**
+   * Return a reference to the object with given name. If the object does
+   * not exist, then the input @p arguments will be used to construct an object
+   * of the given @p Type.
```
Comment (bangerth) [inline include/deal.II/algorithms/general_data_storage.h]: ...a reference to which will then be returned.

same for the previous function

File: include/deal.II/algorithms/general_data_storage.h [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+   */
+  template <typename Type, typename... Args>
+  Type &
+  get_or_add_object_with_name(const std::string &name, Args &&... arguments);
```
Comment (bangerth) [inline include/deal.II/algorithms/general_data_storage.h]: maybe comment here what the difference to the previous function is. (namely, the rvalue references.)

File: include/deal.II/algorithms/general_data_storage.h [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+
+  /**
+   * Return a reference to the object with given name. This function
+   * throws an exception if either an object with the given name is not
+   * stored in this class, or if the object with the given name is not of
+   * compatible type.
```
Comment (bangerth) [inline include/deal.II/algorithms/general_data_storage.h]: ...is not actually of type `Type`

File: include/deal.II/algorithms/general_data_storage.h [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+   */
+  template <typename Type>
+  Type &
+  get_object_with_name(const std::string &name);
+
+  /**
+   * Return a constant reference to the object with the given name. This
+   * function throws an exception if either the name does not exist or if the
+   * object stored in the class with the given name cannot be converted to
+   * the given type.
```
Comment (bangerth) [inline include/deal.II/algorithms/general_data_storage.h]: same here? will any conversions actually be performed?

File: include/deal.II/algorithms/general_data_storage.h [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+
+
+template <class Stream>
+void
+GeneralDataStorage::print_info(Stream &os)
+{
+  for (auto it : any_data)
+    {
+      os << it.first << '\t' << '\t'
+         << boost::core::demangle(it.second.type().name()) << std::endl;
```
Comment (bangerth) [inline include/deal.II/algorithms/general_data_storage.h]: I think it would be useful to explain in the documentation what this function outputs

**Comment by jppelteret** [comment]
@bangerth I think that I've addressed your comments. I've also added a new function `remove_object_with_name`, that does what it says on the box.

File: include/deal.II/algorithms/general_data_storage.h [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+  template <class Stream>
+  void
+  print_info(Stream &stream);
+
+  /**
+   * Clear all data stored in this class instance.
+   *
+   * After this function is called, all copied data owned by this class will
+   * go out of scope. Furthermore, all scoping requirements for data referenced
+   * by this class instance will be lifted.
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/algorithms/general_data_storage.h]: I don't think this was addressed. It's mostly an English language issue related to the grammar you use.

[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/algorithms/general_data_storage.h]: Maybe something like this text:
```
When you call this function, it destroys all objects you asked to be stored as copies, and it forgets about the
references to data you asked to store by reference. As a consequence, you are now free to destroy
the objects to which references were stored at whatever time you want -- before or after the current
`GeneralDataStorage` object is destroyed.
```

**Comment by bangerth** [comment]
I'm curious to next find out where this kind of class is actually useful. I'm not opposed to it, but I suspect that most places where one would want to use it indicates a design flaw. But there's nothing wrong with the class itself, so let's move forward!

/rebuild

**Comment by luca-heltai** [comment]
@jppelteret , it looks like `general_data_storage_01.cc` does not work on the tester. Can you try to figure out what the problem may be?

**Comment by luca-heltai** [comment]
I cannot reproduce the issue. :( 

I have addressed all remaining comments, but I have no write access to @jppelteret branch.

**Comment by luca-heltai** [comment]
A patch that addresses the last comments is here:

949565de6c

**Comment by luca-heltai** [comment]
I think there is still a problem with this. On clang this works:

~~~
data.template get_or_add_object_with_name<std::vector<Number>>(name, n_dofs);
~~~

but apparently gcc does not know how to disambiguate between the `Args &...` and `Args &&...` versions (see the output on jenkins https://jenkins.tjhei.info/blue/organizations/jenkins/dealii/detail/PR-7806/36/pipeline/ for PR #7806).

Do you have an idea on how to solve that?

**Comment by jppelteret** [comment]
Thanks for fixing this @masterleinad ! 👍

**Comment by jppelteret** [comment]
@luca-heltai Do you want to confirm that this works on #7806 before we merge?

**Comment by luca-heltai** [comment]
@masterleinad , I'd have never figured that out... how on earth? :) 

I just forced push #7806. If that succeeds, this can be merged.

**Comment by luca-heltai** [comment]
@masterleinad , would this allow to call 
~~~
auto &value = data.template get_or_add_object_with_name<double>("my_name");
~~~
i.e., without specifying anything at construction?

**Pull request merged**